### ISSUE TITLE
BUG: np.trapz gives (very) different results for integers and floats

### ISSUE THREAD

**Issue by jensdebruijn** [numpy/numpy#24919]
### Describe the issue:

When executing np.trapz negative output values are given when y is given as integer, while output is positive when input is given as float. This only occurs when y values are relatively high. 

### Reproduce the code example:

```python
import numpy as np
np.trapz(x=[0, 1/1000], y=[1771503418, 481833961])
>> -1020814.9585000001

np.trapz(x=[0, 1/1000], y=[1771503418., 481833961.])
>> 1126668.6895

np.trapz(x=[0, 1/1000], y=[20000, 10000])
>> 15.0
```

### Runtime information:

1.24.4
3.11.5 | packaged by conda-forge | (main, Aug 27 2023, 03:23:48) [MSC v.1936 64 bit (AMD64)]

```
[
    {
        "simd_extensions": {
            "baseline": ["SSE", "SSE2", "SSE3"],
            "found": [
                "SSSE3",
                "SSE41",
                "POPCNT",
                "SSE42",
                "AVX",
                "F16C",
                "FMA3",
                "AVX2",
                "AVX512F",
                "AVX512CD",
                "AVX512_SKX",
                "AVX512_CLX",
                "AVX512_CNL",
                "AVX512_ICL"
            ],
            "not_found": []
        }
    },
    {
        "filepath": "C:\\Users\\jbn271\\Miniconda\\envs\\GEB\\Library\\bin\\libblas.dll",
        "internal_api": "mkl",
        "num_threads": 8,
        "prefix": "libblas",
        "threading_layer": "intel",
        "user_api": "blas",
        "version": "2022.1-Product"
    },
    {
        "filepath": "C:\\Users\\jbn271\\Miniconda\\envs\\GEB\\Library\\bin\\libiomp5md.dll",
        "internal_api": "openmp",
        "num_threads": 16,
        "prefix": "libiomp",
        "user_api": "openmp",
        "version": None
    }
]
```

**Comment by JayanthBontha** [comment]
I'm guessing it's due to integer overflow

**Comment by mhvk** [comment]
Indeed, likely an overflow issue: note that on windows the default integer type has only 32 bits! Hence, I don't reproduce this on my Debian machine, but do if I use `y=np.array([1771503418, 481833961], 'i4')`.

This will hopefully be solved in numpy 2.0, when also on windows 64bit integers will be the default. But for now one has explicitly convert to float.

p.s. Closing this, since I think we cannot solve this issue but for defaulting to 64bit, which will happen already. But feel free to re-open if there is more to discuss!
